java/core/src/java/org/apache/orc/impl/SchemaEvolution.java

```diff
@@ -296,13 +303,13 @@ private boolean typesAreImplicitConversion(final TypeDescription fileType,
 
   /**
    * Check if column is safe for ppd evaluation
-   * @param colId reader column id
+   * @param colId file column id
    * @return true if the specified column is safe for ppd evaluation else false
    */
   public boolean isPPDSafeConversion(final int colId) {
     if (hasConversion()) {
-      return !(colId < 0 || colId >= ppdSafeConversion.length) &&
-          ppdSafeConversion[colId];
+      Integer readerTypeId = typeIdsMap.get(colId);
+      return readerTypeId != null && ppdSafeConversion[readerTypeId];
```

Comment:
> Since we only use `ppdSafeConversion[]` in this method, I think changing `ppdSafeConversion[]` to be indexed by file column ids is a simpler solution, i.e. modifying the assignment in `populatePpdSafeConversion` and `populatePpdSafeConversionForChildren` to use file ids.

This looks better. Done